File: archive/cdktf-py/Pr7948/lib/stacks/storage_stack.py
"""Storage infrastructure - S3 buckets with cross-region replication"""

from constructs import Construct
from cdktf_cdktf_provider_aws.s3_bucket import S3Bucket
from cdktf_cdktf_provider_aws.s3_bucket_versioning import S3BucketVersioningA, S3BucketVersioningVersioningConfiguration
from cdktf_cdktf_provider_aws.s3_bucket_server_side_encryption_configuration import (
    S3BucketServerSideEncryptionConfigurationA,
    S3BucketServerSideEncryptionConfigurationRuleA,
    S3BucketServerSideEncryptionConfigurationRuleApplyServerSideEncryptionByDefaultA
)
from cdktf_cdktf_provider_aws.s3_bucket_public_access_block import S3BucketPublicAccessBlock
from cdktf_cdktf_provider_aws.kms_key import KmsKey
from cdktf_cdktf_provider_aws.kms_alias import KmsAlias


class StorageStack(Construct):
    """Creates S3 buckets with cross-region replication and RTC"""

    def __init__(
        self,
        scope: Construct,
        construct_id: str,
        region: str,
        environment_suffix: str,
        is_primary: bool
    ):
        super().__init__(scope, construct_id)

        self.region = region
        self.environment_suffix = environment_suffix
        self.is_primary = is_primary
        self.other_region = "us-east-2" if region == "us-east-1" else "us-east-1"

        # KMS key for S3 encryption
        self.kms_key = KmsKey(
            self,
            "s3-kms-key",
            description=f"KMS key for S3 encryption in {region}",
            deletion_window_in_days=7,
            enable_key_rotation=True,
            tags={
                "Name": f"dr-s3-kms-{region}-{environment_suffix}"
            }
        )

        KmsAlias(
            self,
            "s3-kms-alias",
            name=f"alias/dr-s3-{region}-{environment_suffix}",
            target_key_id=self.kms_key.id
        )

        # S3 Bucket
        self.bucket = S3Bucket(
            self,
            "payment-data-bucket",
            bucket=f"dr-payment-data-{region}-{environment_suffix}",
            force_destroy=True,
            tags={
                "Name": f"dr-payment-data-{region}-{environment_suffix}"
            }
        )

        # Enable versioning (required for replication)
        self.bucket_versioning = S3BucketVersioningA(
            self,
            "bucket-versioning",
            bucket=self.bucket.id,
            versioning_configuration=S3BucketVersioningVersioningConfiguration(
                status="Enabled"
            )
        )

        # Server-side encryption
        S3BucketServerSideEncryptionConfigurationA(
            self,
            "bucket-encryption",
            bucket=self.bucket.id,
            rule=[S3BucketServerSideEncryptionConfigurationRuleA(
                apply_server_side_encryption_by_default=(
                    S3BucketServerSideEncryptionConfigurationRuleApplyServerSideEncryptionByDefaultA(
                        sse_algorithm="aws:kms",
                        kms_master_key_id=self.kms_key.arn
                    )
                ),
                bucket_key_enabled=True
            )]
        )

        # Public access block
        S3BucketPublicAccessBlock(
            self,
            "bucket-public-access-block",
            bucket=self.bucket.id,
            block_public_acls=True,
            block_public_policy=True,
            ignore_public_acls=True,
            restrict_public_buckets=True
        )

        # NOTE: S3 Cross-Region Replication is configured in a separate
        # ReplicationStack that runs after both regional stacks deploy.
        # This ensures the destination bucket exists before replication is configured.

File: archive/cdktf-java/Pr3285/lib/src/main/java/app/stacks/StorageStack.java
package app.stacks;

import com.hashicorp.cdktf.providers.aws.dynamodb_table.DynamodbTable;
import com.hashicorp.cdktf.providers.aws.dynamodb_table.DynamodbTableAttribute;
import com.hashicorp.cdktf.providers.aws.dynamodb_table.DynamodbTablePointInTimeRecovery;
import com.hashicorp.cdktf.providers.aws.dynamodb_table.DynamodbTableServerSideEncryption;
import com.hashicorp.cdktf.providers.aws.kms_alias.KmsAlias;
import com.hashicorp.cdktf.providers.aws.kms_key.KmsKey;
import com.hashicorp.cdktf.providers.aws.s3_bucket.S3Bucket;
import com.hashicorp.cdktf.providers.aws.s3_bucket_public_access_block.S3BucketPublicAccessBlock;
import com.hashicorp.cdktf.providers.aws.s3_bucket_server_side_encryption_configuration.S3BucketServerSideEncryptionConfigurationA;
import com.hashicorp.cdktf.providers.aws.s3_bucket_server_side_encryption_configuration.S3BucketServerSideEncryptionConfigurationRuleA;
import com.hashicorp.cdktf.providers.aws.s3_bucket_server_side_encryption_configuration.S3BucketServerSideEncryptionConfigurationRuleApplyServerSideEncryptionByDefaultA;
import com.hashicorp.cdktf.providers.aws.s3_bucket_versioning.S3BucketVersioningA;
import com.hashicorp.cdktf.providers.aws.s3_bucket_versioning.S3BucketVersioningVersioningConfiguration;
import software.constructs.Construct;

import java.util.Arrays;
import java.util.List;
import java.util.Map;

public class StorageStack {

    private final S3Bucket s3Bucket;

    private final DynamodbTable dynamoTable;

    private final KmsKey s3KmsKey;

    private final KmsKey dynamoKmsKey;

    public StorageStack(final Construct scope, final String id) {
        // Create KMS keys for encryption
        this.s3KmsKey = KmsKey.Builder.create(scope, id + "-s3-kms-key")
                .description("KMS key for S3 bucket encryption")
                .enableKeyRotation(true)
                .tags(Map.of("Name", "s3-encryption-key"))
                .build();

        KmsAlias.Builder.create(scope, id + "-s3-kms-alias")
                .name("alias/s3-serverless-key")
                .targetKeyId(s3KmsKey.getId())
                .build();

        this.dynamoKmsKey = KmsKey.Builder.create(scope, id + "-dynamo-kms-key")
                .description("KMS key for DynamoDB encryption")
                .enableKeyRotation(true)
                .tags(Map.of("Name", "dynamodb-encryption-key"))
                .build();

        // Create S3 bucket with versioning and encryption
        this.s3Bucket = S3Bucket.Builder.create(scope, id + "-data-bucket")
                .bucket("serverless-data-bucket-" + System.currentTimeMillis())
                .tags(Map.of(
                        "Name", "serverless-data-bucket",
                        "Purpose", "Lambda data storage"
                ))
                .build();

        // Enable versioning
        S3BucketVersioningA.Builder.create(scope, id + "-bucket-versioning")
                .bucket(s3Bucket.getId())
                .versioningConfiguration(S3BucketVersioningVersioningConfiguration.builder()
                        .status("Enabled")
                        .build())
                .build();

        // Enable server-side encryption
        S3BucketServerSideEncryptionConfigurationA.Builder.create(scope, id + "-bucket-encryption")
                .bucket(s3Bucket.getId())
                .rule(List.of(
                        S3BucketServerSideEncryptionConfigurationRuleA.builder()
                                .applyServerSideEncryptionByDefault(
                                        S3BucketServerSideEncryptionConfigurationRuleApplyServerSideEncryptionByDefaultA.builder()
                                                .sseAlgorithm("aws:kms")
                                                .kmsMasterKeyId(s3KmsKey.getArn())
                                                .build()
                                )
                                .bucketKeyEnabled(true)
                                .build()
                ))
                .build();

        // Block public access
        S3BucketPublicAccessBlock.Builder.create(scope, id + "-bucket-public-access-block")
                .bucket(s3Bucket.getId())
                .blockPublicAcls(true)
                .blockPublicPolicy(true)
                .ignorePublicAcls(true)
                .restrictPublicBuckets(true)
                .build();

        // Create DynamoDB table
        this.dynamoTable = DynamodbTable.Builder.create(scope, id + "-data-table")
                .name("serverless-data-table")
                .billingMode("PROVISIONED")
                .readCapacity(5)
                .writeCapacity(5)
                .hashKey("pk")
                .rangeKey("sk")
                .attribute(Arrays.asList(
                        DynamodbTableAttribute.builder()
                                .name("pk")
                                .type("S")
                                .build(),
                        DynamodbTableAttribute.builder()
                                .name("sk")
                                .type("S")
                                .build()
                ))
                .serverSideEncryption(DynamodbTableServerSideEncryption.builder()
                        .enabled(true)
                        .kmsKeyArn(dynamoKmsKey.getArn())
                        .build())
                .pointInTimeRecovery(DynamodbTablePointInTimeRecovery.builder()
                        .enabled(true)
                        .build())
                .tags(Map.of(
                        "Name", "serverless-data-table",
                        "Purpose", "Application data storage"
                ))
                .build();
    }

    // Getters
    public S3Bucket getS3Bucket() {
        return s3Bucket;
    }

    public DynamodbTable getDynamoTable() {
        return dynamoTable;
    }

    public KmsKey getS3KmsKey() {
        return s3KmsKey;
    }

    public KmsKey getDynamoKmsKey() {
        return dynamoKmsKey;
    }
}
